### PR TITLE
Add new splat mode, add hiding support

### PR DESCRIPTION
**Pull request by slimbuck**
Fixes: #50, #29, #47

This PR:
* introduces a new rendering/selection mode called "Rings", which allows selection by splat extents instead of splat centers
* adds an option to hide selected splats and unhide them
* adds a new picker selection tool to extend rect and brush
* uses a state array to track each splat state, instead of updating splat opacity
* double click the scene to center the orbit camera around the target point
* update npm dependencies, including engine [v1.70.0](https://github.com/playcanvas/engine/releases/tag/v1.70.0)

Screenshot 2024-04-19 at 12 01 12

### CODE REVIEW

File: src/camera.ts [hunk 1]
```diff
@@ -259,17 +293,29 @@ class Camera extends Element {
 
         // in with the new
         const pixelFormat = PIXELFORMAT_RGBA8;
+        const samples = this.scene.config.camera.multisample ? device.maxSamples : 1;
```
Comment (mvaligursky) [inline src/camera.ts:296]: any benefits to use multisampling in this tool?

File: src/camera.ts [hunk 2]
```diff
@@ -259,17 +293,29 @@ class Camera extends Element {
...
         this.entity.camera.renderTarget = renderTarget;
         this.entity.camera.camera.horizontalFov = width < height;
+
+        // create pick mode render target
+        this.pickModeColorBuffer = createTexture(width, height, pixelFormat);
+        this.pickModeRenderTarget = new RenderTarget({
+            colorBuffer: this.pickModeColorBuffer,
+            depth: false,
+            flipY: false,
+            samples: samples,
```
Comment (mvaligursky) [inline src/camera.ts:316]: I assume the picker rendering here renders some IDs. Is it compatible with multisampling? I assume the resolve would average IDs.

File: src/editor.ts [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+            const state = splatData.getProp('state') as Uint8Array;
+            const x = splatData.getProp('x');
+            const y = splatData.getProp('y');
+            const z = splatData.getProp('z');
+
+            vec.set(axis[0], axis[1], axis[2]);
+            vec2.set(axis[0] * distance, axis[1] * distance, axis[2] * distance);
+
+            // transform the plane to local space
+            mat.invert(splatDef.element.worldTransform);
```
Comment (mvaligursky) [inline src/editor.ts:401]: Not entirely sure what matrix gets inverted here .. just making sure this needs to take place in the loop

**Review by mvaligursky — APPROVED** [review]
Awesome!

File: src/camera.ts (hunk 1 above)
Comment (slimbuck) [reply src/camera.ts:296]: nope, it's disabled in config https://github.com/playcanvas/super-splat/blob/main/src/scene-config.ts#L23, but I decided to keep support in the code for now.

File: src/camera.ts (hunk 2 above)
Comment (slimbuck) [reply src/camera.ts:316]: probably not compatible with resolve, but as mentioned above it's disabled. i might completely remove multisampling support in the next update.

File: src/editor.ts (hunk 3 above)
Comment (slimbuck) [reply src/editor.ts:401]: The loop here is over instances of GSplat in the scene, not individual splats. The loop over splats happens in `processSelection` below. 👍🏻

File: src/splat.ts [hunk 4]
```diff
@@ -81,13 +204,46 @@ class Splat extends Element {
...
-        const opacity = this.splatData.getProp('opacity');
-        this.splatData.calcAabb(localBound, (i: number) => {
-            return opacity[i] > -1000;
-        });
+        localBound.center.set((minx + maxx) * 0.5, (miny + maxy) * 0.5, (minz + maxz) * 0.5);
+        localBound.halfExtents.set((maxx - minx) * 0.5, (maxy - miny) * 0.5, (maxz - minz) * 0.5);
+
+        // this.splatData.calcAabb(localBound, (i: number) => {
+        //     return opacity[i] > -1000;
+        // });
```
Comment (willeastcott) [inline src/splat.ts:246]: Delete?

File: src/tools/picker-selection.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+class PickerSelection {
+    events: Events;
+    root: HTMLElement;
+
+    constructor(events: Events, parent: HTMLElement) {
+        this.root = document.createElement('div');
+        this.root.id = 'select-root';
+
+        this.root.onmousedown = (e: MouseEvent) => {
```
Comment (willeastcott) [inline src/tools/picker-selection.ts:11]: Can we aim to stick to pointer events in our apps from now on?

**Review by willeastcott — APPROVED** [review]
Can't wait for this to be released.

File: src/tools/picker-selection.ts (hunk 5 above)
Comment (slimbuck) [reply src/tools/picker-selection.ts:11]: Yes! I will update everything to pointer events in a separate PR.

**Pull request merged**